ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoMaterialProperties.cpp

```diff
@@ -0,0 +1,294 @@
+    return _viscosity->getValue(vars);
+}
+
+double TwoPhaseFlowWithPrhoMaterialProperties::getGasViscosity(
+    const double p, const double T) const
+{
+    ArrayType vars;
+    vars[static_cast<int>(MaterialLib::Fluid::PropertyVariableType::T)] = T;
+    vars[static_cast<int>(MaterialLib::Fluid::PropertyVariableType::p)] = p;
+    return _gas_viscosity->getValue(vars);
+}
+
+Eigen::MatrixXd const& TwoPhaseFlowWithPrhoMaterialProperties::getPermeability(
+    const double /*t*/, const ProcessLib::SpatialPosition& /*pos*/,
+    const int /*dim*/) const
+{
+    return _intrinsic_permeability_models[_current_material_id];
+}
+
+double TwoPhaseFlowWithPrhoMaterialProperties::getPorosity(
+    const double /*t*/, const ProcessLib::SpatialPosition& /*pos*/,
+    const double /*p*/, const double T, const double porosity_variable) const
+{
+    const double porosity =
+        _porosity_models[_current_material_id]->getValue(porosity_variable, T);
+
+    return porosity;
+}
+
+double TwoPhaseFlowWithPrhoMaterialProperties::getNonwetRelativePermeability(
+    const double /*t*/, const ProcessLib::SpatialPosition& /*pos*/,
+    const double /*p*/, const double /*T*/, const double saturation) const
+{
+    const double nonwet_krel =
+        _relative_permeability_models[0]->getValue(saturation);
+
+    return nonwet_krel;
+}
+
+double TwoPhaseFlowWithPrhoMaterialProperties::getWetRelativePermeability(
+    const double /*t*/, const ProcessLib::SpatialPosition& /*pos*/,
+    const double /*p*/, const double /*T*/, const double saturation) const
+{
+    const double wet_krel =
+        _relative_permeability_models[1]->getValue(saturation);
+
+    return wet_krel;
+}
+
+double TwoPhaseFlowWithPrhoMaterialProperties::getCapillaryPressure(
+    const double /*t*/, const ProcessLib::SpatialPosition& /*pos*/,
+    const double /*p*/, const double /*T*/, const double saturation) const
+{
+    const double pc =
+        _capillary_pressure_models[_current_material_id]->getCapillaryPressure(
+            saturation);
+    return pc;
+}
+
+double TwoPhaseFlowWithPrhoMaterialProperties::getDerivCapillaryPressure(
+    const double /*t*/, const ProcessLib::SpatialPosition& /*pos*/,
+    const double /*p*/, const double /*T*/, const double saturation) const
+{
+    const double dpcdsw =
+        _capillary_pressure_models[_current_material_id]->getdPcdS(saturation);
+    return dpcdsw;
+}
+
+bool TwoPhaseFlowWithPrhoMaterialProperties::computeConstitutiveRelation(
+    double const t,
+    ProcessLib::SpatialPosition const& x,
+    double const PG,
+    double const X,
+    double& Sw,
+    double& X_m,
+    double& dsw_dpg,
+    double& dsw_dX,
+    double& dxm_dpg,
+    double& dxm_dX)
+{
+    using LocalJacobianMatrix = Eigen::Matrix<double, 2, 2, Eigen::RowMajor>;
+    LocalJacobianMatrix J_loc;
+    // Linear solver for the newton loop is required after the loop with the
+    // same matrix. This saves one decomposition.
```

Comment:
You say the linear solver is required after the newton loop, but I could not find it.